xVHD - Add support for setting disk type (Fixed, Dynamic, Differencing)

johnf opened the pull request:
Add the ability to create Fixed disks

Fixes #72 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xhyper-v/77)
<!-- Reviewable:end -->

msftclas commented:
Hi __@johnf__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

PlagueHO commented:
Hi @johnf - are you able to fix the problem in the MOF schema as reported by the AppVeyor tests failing?

johnf commented:
Fixed. Dumb typo

PlagueHO commented:
@johnf - awesome job! Thanks for addressing this.

Could you please add an entry to the #Unreleased section in the Readme.md listing the change?

Could you also add this to this parameter to the Examples/Sample_xVHD_DiffVHD.ps1 (and to the same place in the Examples section of the Readme.md).

It might be worth also creating an Examples/Sample_xVHD_FixedVHD.ps1 showing how the resource is used to create a fixed disk (and copy the example to the Readme.md)

Unfortunately this resource doesn't contain any unit tests for xVHD yet so we can't add tests for this change.

Thanks again for all your work on this! It is greatly appreciated.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 65 at r1](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ka0t85oQd-PcDyqG8w8:-Ka0t85piG7cVnLjw2U5:b-ix0fsb) ([raw file](https://github.com/powershell/xhyper-v/blob/b6f8eb35b4bc0ef2e9aa1b066e084920cd7154d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L65)):*
> ```PowerShell
>         # Type of Vhd to be created
>         [ValidateSet("Dynamic","Fixed","Differencing")]
>         [String]$Type = "Dynamic",
> ```

One thing that may be confusing here is that it would be possible for someone to specify a $Type of 'Dynamic' or 'Fixed' and also a $ParentPath set. This doesn't look like it would cause any problems, but it might imply to users that this was a possible combination of parameters (e.g. a Dynamic Differencing Disk- which isn't of course possible).

It might be worth adding a parameter validation check into the Set-* and Test-* that would throw an exception if an invalid combination of parameters was requested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-Ka0qZHwjjxZb6TtQ7ER:b-kbxv55)*
<!-- Sent from Reviewable.io -->

johnf commented:
I've updated my branch with your comments.

I added the new Example but in the README rather than add a whole new example I updated the first one to be Fixed.

bgelens commented:
@johnf We have diverged a bit with the Readme since this PR was created. Examples are no longer part of it. Instead links are added to the example in the Examples folder. Probably a good idea to rebase and address the comments so we can take the PR.

Sorry to put you through this. Let me know if you are not comfortable with concepts as rebasing and such.

---

Reviewed 1 of 3 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 64 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8vajBXYWprUj5YgJM:-Ko8vajCxNmrYLXBkUOC:by02azk) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L64)):*
> ```PowerShell
> 
>         # Type of Vhd to be created
>         [ValidateSet("Dynamic","Fixed","Differencing")]
> ```

Could you add ```[Parameter()]``` on top of the ValidateSet and Place $Type on a new line below ```[String]```

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 83 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8ullhIEw46tsxQTO0:-Ko8ullhIEw46tsxQTO1:b-nit6vt) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L83)):*
> ```PowerShell
> 
>     # Check the module combinations make sense
>     if($ParentPath -and $Type -ne "Differencing")
> ```

You'll also need something like if  ```($null -eq $parentpath -and $Type -eq 'Differencing'){throw error}```

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 151 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8w7XodMbtEDC1tasR:-Ko8w7XodMbtEDC1tasS:bkli07r) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L151)):*
> ```PowerShell
>                 if($vhd.Type -ne $Type)
>                 {
>                     Throw "This module can't convert disk types yet"
> ```

Are you planning to implement this, if not, please remove ```yet``` :smile:

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 205 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8wFDYyVF3gjLnRFqt:-Ko8wFDYyVF3gjLnRFqu:b-twta97) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L205)):*
> ```PowerShell
> 
>         # Type of Vhd to be created
>         [ValidateSet("Dynamic","Fixed","Differencing")]
> ```

Same comments as with Set-TargetResource apply here

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 227 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8wLeNHrDILhqUeDfm:-Ko8wLeNHrDILhqUeDfn:bkkjs1q) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L227)):*
> ```PowerShell
> 
>     # Check the module combinations make sense
>     if($ParentPath -and $Type -ne "Differencing")
> ```

And here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-Ko8x2o3sapq4Jde8CTH:b9x0xkl)*
<!-- Sent from Reviewable.io -->

johnf commented:
I've rebased

bgelens commented on DSCResources/MSFT_xVHD/MSFT_xVHD.psm1:
```diff
@@ -75,6 +80,17 @@ function Set-TargetResource
     }
 
+    # Check the module combinations make sense
+    if($ParentPath -and $Type -ne "Differencing")
```
Could you add a unit test for this resource. It doesn't have to cover the entire thing but at least validate these parameter conditions are working properly e.g. It 'Should throw when ParentPath defined and Type -ne Diff'

bgelens commented on DSCResources/MSFT_xVHD/MSFT_xVHD.psm1:
```diff
@@ -60,6 +60,11 @@ function Set-TargetResource
         [Uint64]$MaximumSizeBytes,
 
+        # Type of Vhd to be created
+        [parameter()]
```
Should be:
```powershell
[Parameter()]
[ValidateSet('Dynamic', 'Fixed', 'Differencing')]
[String]
$Type = 'Dynamic',
```

bgelens commented on DSCResources/MSFT_xVHD/MSFT_xVHD.psm1:
```diff
@@ -75,6 +80,17 @@ function Set-TargetResource
+
+    if($null -eq $ParentPath -and $Type -eq "Differencing")
+    {
+        Throw "Differencing requires a parent path"
```
And It 'Should throw when ParentPath is not defined and type is Differencing'

bgelens commented on DSCResources/MSFT_xVHD/MSFT_xVHD.psm1:
```diff
@@ -181,6 +207,11 @@ function Test-TargetResource
         [String]$Generation = "Vhd",
 
+        # Type of Vhd to be created
+        [parameter()]
```
Same as with other comment

bgelens commented on DSCResources/MSFT_xVHD/MSFT_xVHD.psm1:
```diff
@@ -198,7 +229,18 @@ function Test-TargetResource
     }
-            
+
+    # Check the module combinations make sense
```
Could you remove this comment or move it up (so the earlier validation is included under the comment as well) and change the wording so that it says something like, Validating Parameters

johnf commented:
I written some tests which are close but aren't working quite right.
I'm not a powershell programmer and I don't have an environment at the moment where I can actually run the tests.
And the cycle of doing that via appveyor will take me hours :)

I've ticked ```allow edits from maintainers``` so you can commit to my branch if someone else can make them work

bgelens commented:
@johnf I've send in some commits to your repo to make some adjustments.
@PlagueHO could you review this PR (as I contributed the last commits, I would like some other eyes looking at the changes) The added unit tests are just an initial base set for parameter combination validation. Additional tests need to be added and I created an issue to increase code coverage already #119 )

---

Reviewed 2 of 3 files at r3, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoJINA3VEemzgwXLd-i:b-82df9e)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Sorry about the minor nitpicks. Generally things look fine, but if you're going to do some rework on this module at some point then I've made some suggested changes. One other thing that you would want to address in any rework: Move strings to strings file.

---

Reviewed 1 of 4 files at r2, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 83 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-Ko8ullhIEw46tsxQTO0:-KoTdR-2Au3x8kD3xorK:branein) ([raw file](https://github.com/powershell/xhyper-v/blob/f3b2f3ee13d9a55b64ac6b51ff085cdedd55560f/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L83)):*
<details><summary><i>Previously, bgelens (Ben Gelens) wrote…</i></summary><blockquote>

You'll also need something like if  ```($null -eq $parentpath -and $Type -eq 'Differencing'){throw error}```
</blockquote></details>

I'd generally recommend creating an `Assert-ParametersValid` function in a resource and using that to validate the parameter input - with throwing if they are invalid. But don't worry about doing that here. Just a suggestion for future.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 233 at r3](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-K_wLGx1sDmn2Lv3dt5b-r3-233:-KoTf2neNceuZUDVb-cj:b-jszjx8) ([raw file](https://github.com/powershell/xhyper-v/blob/1eb9ae4187ff2b6790c6d8c5afa6552e0b8d8cca/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L233)):*
<details><summary><i>Previously, bgelens (Ben Gelens) wrote…</i></summary><blockquote>

Could you remove this comment or move it up (so the earlier validation is included under the comment as well) and change the wording so that it says something like, Validating Parameters
</blockquote></details>

See my earlier comments about parameter validation in `Assert-ParametersValid` - again not at all required here. Just if anyone is doing rework on this module that would be the recommended pattern.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 37 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTgNebIdz3tra6kuUq:-KoTgNecYJhU0WTwpLfx:b-oacgyv) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L37)):*
> ```PowerShell
>     if(!(Get-Module -ListAvailable -Name Hyper-V))
>     {
>         Throw 'Please ensure that Hyper-V role is installed with its PowerShell module'
> ```

Not required here (so please don't worry about fixing in this PR), but just a suggestion: convert to standard way of throwing custom errors. E.g. New-InvalidOperationException.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 137 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTeKg59miHIUPwJx-n:-KoTeKg6KczsAicirLjB:b-453w6q) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L137)):*
> ```PowerShell
> 
>     # If vhd should be absent, delete it
>     if($Ensure -eq 'Absent')
> ```

Nitpick: Need space between if and ( - just use global search/replace to fix them all.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 197 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTe36QzDKux2a9sZxj:-KoTe36RUx65iSJezZVI:b-sfp5yf) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L197)):*
> ```PowerShell
> 
>         # Vhd file is not present
>         catch
> ```

Hope I'm not being too much of a pain, but won't this catch any kind of error - not just missing VHD files?

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 212 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTeSkaZOrxE1bSMZea:-KoTeSkaZOrxE1bSMZeb:b-xnum63) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L212)):*
> ```PowerShell
>                 }
>                 $null = New-VHD @params
>             }
> ```

Please add blank line after end of code blocks.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 282 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTeh0frVS_sjSpSinI:-KoTeh0gzGWlSymkzvPb:boyh6yp) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L282)):*
> ```PowerShell
> 
>     # Check if Hyper-V module is present for Hyper-V cmdlets
>     if(!(Get-Module -ListAvailable -Name Hyper-V))
> ```

Not required, please don't worry about doing this now, just a suggestion for future: The recommended way of doing this is to move it to an `Assert-HyperVRequired` function.

---

*[Examples/Sample_xVHD_FixedVHD.ps1, line 5 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTgrvVHzixnQgkneHt:-KoTgrvVHzixnQgkneHu:b-4gi2f5) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Examples/Sample_xVHD_FixedVHD.ps1#L5)):*
> ```PowerShell
>     param
>     (
>         [string[]]$NodeName = 'localhost',
> ```

Please convert parameter block to be standard format. E.g.
```powershell
[Parameter()]
[String[]]
$NodeName = 'localhost',

[Parameter(Mandatory = $true)]
[String]
$Name
```

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 47 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTfW-OGGArRm8kev5y:-KoTfW-OGGArRm8kev5z:b-trhyla) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L47)):*
> ```PowerShell
>                     return $false
>                 }
>                 It 'Should throw when the module is missing' {
> ```

Add blank line here.

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 126 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoThDuTzF1oJ3cNFq4O:-KoThDuU7j4ORDidERxe:b-749uj6) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L126)):*
> ```PowerShell
>                 Mock -CommandName Get-Module -ParameterFilter { ($Name -eq 'Hyper-V') -and ($ListAvailable -eq $true) } -MockWith {
>                     return $false
>                 }
> ```

Please add blank line here.

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 162 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTff4MPU4BiPQh62Fd:-KoTff4MPU4BiPQh62Fe:btq66kg) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L162)):*
> ```PowerShell
>                 }
> 
>                 #"Generation $Generation should match ParentPath extension $($ParentPath.Split('.')[-1])"
> ```

Space after #

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 164 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoThHwEx7C5YS_3SZWv:-KoThHwEx7C5YS_3SZWw:b-749uj6) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L164)):*
> ```PowerShell
>                 #"Generation $Generation should match ParentPath extension $($ParentPath.Split('.')[-1])"
>                 Mock -CommandName Test-Path -MockWith { $true }
>                 It 'Should throw when file extension and generation have a mismatch' {
> ```

Please add blank line here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoThTLomYRi4XSYEhpX:b8bh23s)*
<!-- Sent from Reviewable.io -->

bgelens commented:
No worries, keep up with the nitpicks!

I've made some changes per your findings. Please approve these if they are to your liking :smile:

I'm not going to improve / enhance this resource further within the scope of this PR. The additions I made to it, felt needed to have this PR merged (most important: adding the test cases which resulted in some of the code being changed due to bug findings :wink: )
So further enhancements need to be addressed in a new PR. I've logged an issue #121 which should kick start the work on this.

---

Reviewed 2 of 2 files at r6.
Review status: 3 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 137 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTeKg59miHIUPwJx-n:-KoW9qbSigMzI0O0XENF:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L137)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Nitpick: Need space between if and ( - just use global search/replace to fix them all.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xVHD/MSFT_xVHD.psm1, line 197 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTe36QzDKux2a9sZxj:-KoWAS9Uiqgqzqz8W9lC:b-iegacj) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/DSCResources/MSFT_xVHD/MSFT_xVHD.psm1#L197)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Hope I'm not being too much of a pain, but won't this catch any kind of error - not just missing VHD files?
</blockquote></details>

You are not, it is true and I was not happy with this myself. I already had made some changes to this PR original contributors code as this was actually the case with the type mismatch .
I'm not planning to reevaluate this resource in this PR but I've logged an issue #121

---

*[Examples/Sample_xVHD_FixedVHD.ps1, line 5 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTgrvVHzixnQgkneHt:-KoWCouH34yp0ogIl7SH:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Examples/Sample_xVHD_FixedVHD.ps1#L5)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Please convert parameter block to be standard format. E.g.
```powershell
[Parameter()]
[String[]]
$NodeName = 'localhost',

[Parameter(Mandatory = $true)]
[String]
$Name
```
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 47 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTfW-OGGArRm8kev5y:-KoWE9Z7G2kQBDEtX5Sf:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L47)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Add blank line here.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 126 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoThDuTzF1oJ3cNFq4O:-KoWEFQLNaUOab7oBVcg:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L126)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Please add blank line here.
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 162 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoTff4MPU4BiPQh62Fd:-KoWEQs-bxeZsG2zbM1t:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L162)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Space after #
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xVHD.tests.ps1, line 164 at r6](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoThHwEx7C5YS_3SZWv:-KoWEdxAT3OOcPPnaYFZ:bb6un1j) ([raw file](https://github.com/powershell/xhyper-v/blob/bb5ac47764d6f1efea17f03e7f9fad6f1bd7b7d6/Tests/Unit/MSFT_xVHD.tests.ps1#L164)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Please add blank line here.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoWH0Kr4kbrdPCSSQdX:b-jnje9u)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Thanks for all your work on getting this PR out :grin: Nearly there! You're doing great stuff.

---

Reviewed 1 of 4 files at r2, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Examples/Sample_xVHD_DiffVHD.ps1, line 19 at r7](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoYqTDUItktdT5ejrHt:-KoYqTDVn6HcDYlNdCeC:b-l5h9n5) ([raw file](https://github.com/powershell/xhyper-v/blob/1da8a251ae33f46a7dc5acb52281221f202b2425/Examples/Sample_xVHD_DiffVHD.ps1#L19)):*
> ```PowerShell
>         [string]$Generation = "Vhd",
> 
>         [ValidateSet("Dynamic","Fixed","Differencing")]
> ```

This file looks like it still needs it's parameter style fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoYqgbWAAkzimUZjwWE:b-iulito)*
<!-- Sent from Reviewable.io -->

bgelens commented:
OK please check again :wink:

---

Reviewed 4 of 4 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Examples/Sample_xVHD_DiffVHD.ps1, line 19 at r7](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-KoYqTDUItktdT5ejrHt:-KoZ4RxmIWPCvGpTtFft:byd0dn7) ([raw file](https://github.com/powershell/xhyper-v/blob/1da8a251ae33f46a7dc5acb52281221f202b2425/Examples/Sample_xVHD_DiffVHD.ps1#L19)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

This file looks like it still needs it's parameter style fixed.
</blockquote></details>

Done plus all other xVHD examples

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoZ4a0N1aC-zqpion7B:bvwlfds)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Great job and done!

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r8.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/77#-:-KoZFgXRWE5QnaONUFxa:b-mgvkdz)*
<!-- Sent from Reviewable.io -->
